usr/src/uts/common/fs/zfs/dsl_dataset.c

```diff
@@ -883,14 +884,41 @@ dsl_dataset_create_sync_dd(dsl_dir_t *dd, dsl_dataset_t *origin,
 	return (dsobj);
 }
 
+static int
+deadlist_enqueue_cb(void *arg, const blkptr_t *bp, dmu_tx_t *tx)
```

Comment:
It kind of sucks to have the same exact function defined in 2 files (here and dsl_pool.c), even though it is pretty trivial. Would you consider adding this function:

```
void
dsl_dataset_sync_done(dsl_dataset_t *ds, dmu_tx_t *tx)
{
	objset_t *os = ds->ds_objset;

	bplist_iterate(&ds->ds_pending_deadlist,
	    deadlist_enqueue_cb, &ds->ds_deadlist, tx);

	ASSERT(!dmu_objset_is_dirty(os, dmu_tx_get_txg(tx)));

	dmu_buf_rele(ds->ds_dbuf, ds);
}
```

And then calling it from `dsl_pool_sync()`, that way we can get rid of the `deadlist_enqueue_cb()` in `dsl_pool.c`.